snakemake/executors/flux.py

```diff
@@ -0,0 +1,236 @@
+__author__ = "Johannes Köster"
+__copyright__ = "Copyright 2022, Johannes Köster"
+__email__ = "johannes.koester@uni-due.de"
+__license__ = "MIT"
+
+import os
+import shlex
+import sys
+from collections import namedtuple
+
+from snakemake.executors import ClusterExecutor, sleep
+from snakemake.executors.common import format_cli_arg, join_cli_args
+from snakemake.logging import logger
+from snakemake.resources import DefaultResources
+
+# Just import flux once
+try:
+    import flux
+    import flux.job
+except ImportError:
+    flux = None
+
+FluxJob = namedtuple("FluxJob", "job jobname jobid callback error_callback flux_future")
+
+
+class FluxExecutor(ClusterExecutor):
+    """
+    The Flux executor deploys workflows to a flux cluster.
+    """
+
+    def __init__(
+        self,
+        workflow,
+        dag,
+        cores,
+        jobname="snakejob.{name}.{jobid}.sh",
+        printreason=False,
+        quiet=False,
+        printshellcmds=False,
+    ):
+        super().__init__(
+            workflow,
+            dag,
+            None,
+            jobname=jobname,
+            printreason=printreason,
+            quiet=quiet,
+            printshellcmds=printshellcmds,
+            assume_shared_fs=False,
+            max_status_checks_per_second=10,
+        )
+
+        # Attach variables for easy access
+        self.workdir = os.path.realpath(os.path.dirname(self.workflow.persistence.path))
+        self.envvars = list(self.workflow.envvars) or []
+
+        # Quit early if we can't access the flux api
+        if not flux:
+            raise WorkflowError(
+                "Cannot import flux. Is a cluster available to you with Python bindings?"
```

Comment:
The way Flux is designed now, you usually get the core software installed with the Python bindings - and we don't yet have a "go off and install flux Python" on your own sort of thing. So the expected thing is that they would need to get access to a cluster, or ask their HPC admin. I can put a link to the GitHub repository or docs if you like, but I'm not sure it's going to be useful to most users. But I'm not confident in my answer - ping @trws are there cases where users are able/willing to install their own Flux?

Reply:
FYI - Since Flux can run under other resource managers or very easily standalone with `flux start` it is much more likely users might install their own copy (or use the spack package) than a traditional RM.

Reply:
Hmm so maybe we should direct them to an install page. @grondo which one would you suggest?

Reply:
If the current cluster is running Flux and the bindings can't be imported then most of the Flux commands won't work so users will already be contacting their sysadmins. I'd just point users to the homepage on error as suggested.

Reply:
okay added! https://github.com/snakemake/snakemake/pull/1810/commits/92d78e827cf6753669a6f41211f5b0e01f4f3a00